system/jlib/jsecrets.cpp

```diff
@@ -249,59 +259,171 @@ class CVault
-                token.set(tokenText.str());
+                if (getSecretKeyValue(clientToken, clientSecret, "token"))
+                {
+                    authType = VaultAuthType::appRole;
+                    PROGLOG("using a client token for vault auth");
+                }
             }
         }
+        else if (isContainerized())
+        {
+            authType = VaultAuthType::k8s;
+            if (vault->hasProp("@role"))
+                k8sAuthRole.set(vault->queryProp("@role"));
+            else
+                k8sAuthRole.set("hpcc-vault-access");
+            DBGLOG("using kubernetes vault auth");
+        }
+    }
+    inline const char *queryAuthType()
+    {
+        switch (authType)
+        {
+            case VaultAuthType::appRole:
+                return "approle";
+            case VaultAuthType::k8s:
+                return "kubernetes";
+            case VaultAuthType::token:
+                return "token";
+        }
+        return "unknown";
+    }
+    void vaultAuthError(const char *msg)
+    {
+        Owned<IException> e = makeStringExceptionV(0, "Vault [%s] %s auth error %s", name.str(), queryAuthType(), msg);
+        OERRLOG(e);
+        throw e.getClear();
     }
+    void vaultAuthErrorV(const char* format, ...) __attribute__((format(printf, 2, 3)))
+    {
+        va_list args;
+        va_start(args, format);
+        StringBuffer msg;
+        msg.valist_appendf(format, args);
+        va_end(args);
+        vaultAuthError(msg);
+    }
+    void processClientTokenResponse(httplib::Result &res)
+    {
+        if (!res)
+            vaultAuthError("missing login response");
+        if (res->status != 200)
+            vaultAuthErrorV("[%d](%d) - response: %s", res->status, res.error(), res->body.c_str());
+        const char *json = res->body.c_str();
+        if (isEmptyString(json))
+            vaultAuthError("empty login response");
+
+        Owned<IPropertyTree> respTree = createPTreeFromJSONString(json);
+        if (!respTree)
+            vaultAuthError("parsing JSON response");
+        const char *token = respTree->queryProp("auth/client_token");
+        if (isEmptyString(token))
+            vaultAuthError("response missing client_token");
+
+        clientToken.set(token);
+        clientTokenRenewable = respTree->getPropBool("auth/renewable");
+        unsigned lease_duration = respTree->getPropInt("auth/lease_duration");
+        if (lease_duration==0)
+            clientTokenExpiration = 0;
+        else
+        {
+            CDateTime t;
+            t.setNow();
+            t.adjustTimeSecs(lease_duration);
+            clientTokenExpiration = t.getSimple();
+        }
+        DBGLOG("VAULT TOKEN duration=%d", lease_duration);
+    }
+    bool isClientTokenExpired()
+    {
+        if (clientTokenExpiration==0)
+            return false;
+        double remaining = difftime(clientTokenExpiration, time(nullptr));
```

Comment:
Consistency. This uses time,  other code uses CDateTime.  It isn't immediately clear they are compatible

Reply:
@ghalliday When using CDateTime I was actually trying to avoid the assumption that time_t was stored as seconds (for non posix systems?),.. but I see we make that assumption all over the place, including inside CDateTime.  I'll use time() consistently instead.